src/avram/postgres_url.cr

```diff
@@ -1,39 +1,72 @@
 class Avram::PostgresURL
-  getter database, username, password, hostname, port
+  getter database, username, password, hostname, port, query
+  property url : String = ""
 
   def initialize(
     @database : String,
-    @hostname : String,
+    @hostname : String = "",
     @username : String = "",
     @password : String = "",
-    @port : String = ""
+    @port : String = "",
+    @query : String = ""
   )
   end
 
-  def self.build(**args)
+  def self.build(**args) : PostgresURL
     new(**args).build
   end
 
+  def self.parse(nothing : Nil)
+    nil
+  end
+
+  def self.parse(url : String) : PostgresURL
+    uri = URI.parse(url)
+    build(
+      database: uri.path.to_s,
+      hostname: uri.host.to_s,
+      username: uri.user.to_s,
+      password: uri.password.to_s,
+      port: uri.port.to_s,
+      query: uri.query.to_s
+    )
+  end
+
   def build
-    String.build do |io|
+    sanitize_inputs
+    self.url = String.build do |io|
       set_url_protocol(io)
       set_url_creds(io)
       set_url_host(io)
       set_url_port(io)
       set_url_db(io)
     end
+    self
+  end
+
+  private def sanitize_inputs
```

Comment:
The code I write, the more I want simple in-and-out (pure) functions. Less room for unexpected values and strange mutation bugs. How do you feel about moving each of these into their `set_url_<part>` method and taking care of `strip`ing or doing whatever needs to be done there? It also stops this method from doing double duty (sanitizing and validating the `@database` value).